### DISCUSSION TITLE
Can i join the cpp and lua templates?

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1590] [General]
Hi,

Can i join the cpp and lua templates?

I need to know if it's in your interest before making the pull-request and then getting rejected and investing time in it.

I can do?

**Comment by aismann** [comment]
> Can i join the cpp and lua templates?
> I need to know if it's in your interest before making the pull-request and then getting rejected and investing time in it.
> I can do?

As @rh101 has written sometimes before:
```
We have to be careful when making changes to the template. 
The template should remain as simple as possible, and not just adds "noise" or your own stuff to the code.
```

And:
```
The template is meant as a starting point, and that file will be edited during your development process. 
The developer of a project can easily add or delete any code they want to it.
```

**Comment by iAndyHD3** [comment]
Say what you would like to change so it can be discussed first to avoid wasted effort

**Comment by rh101** [comment]
While the templates do have similarities, they are not the same.  The LUA template has a lot of modifications and extra files that are specific to LUA, and would cause problems with projects not using LUA.  

Merging them together would add a lot of work for developers who do not use LUA, forcing them to clean up the generated project before they can even start their own work.  

It makes no sense to merge them whatsoever.

**Comment by paulocoutinhox** [comment]
Hi,

I said that i want join the template, don't change anything.

I think that will be more easy keep a single template, since they share most things.

The only files that have lua specific things is "AppDelegate.cpp" and "CMakeLists.txt" that i can have two and keep only one after project generation.

Files not used by selected template will be automatically removed, like it is done today inside "axproj-template.json", it solve the problem of have only needly files for that template.

**Reply by aismann** [reply]
>I think that will be more easy keep a single template, since they share most things.

What about the future? 
Will C++/Lua template be the same?
I think the improvement will be not so much.

**Reply by rh101** [reply]
If you can ensure that the CPP projects generated from that single template will be exactly as they are now (meaning no LUA related items in the CPP project), then yes, it would be reasonable to merge them.  You would also have to ensure that merging them isn't going to add an extra level of complexity to maintaining the template.

**Reply by paulocoutinhox** [reply]
Merging projects does not mean that they will have both codes in the final project generated by the tool, but that we will have a single template to maintain. Obviously, the final code will be what we have today, but I don't need to say that this will be the case, right?

**Reply by rh101** [reply]
> Obviously, the final code will be what we have today, but I don't need to say that this will be the case, right?

No, everything should always be stated plainly.  If a proposal is put forward to make changes, then it is always best to be clear about why the changes are being made, and what impact they may have.  In this case, stating that the modifications to the template would not impact the generated projects may have been enough to avoid all of this back and forth discussion.

**Comment by aismann** [comment]
But anyhow:
@paulocoutinhox 
Let see what you want improve (not change!)

**Reply by paulocoutinhox** [reply]
What we will gain from this is in terms of maintenance. Today, people only change the cpp-template code and not the Lua template, leaving the codes different. We have to change/add images for the two templates that have the same images and xcode things, for example. Basically our work is doubled, and can be reduced to a single job, and this is something you haven't realized yet. If we only have one place to tinker, the rework is cut out and we can move forward better by tinkering with a single project.

**Reply by aismann** [reply]
> and this is something you haven't realized yet

@paulocoutinhox 
Belive me, we known it. Make a PR and let see.

**Comment by aismann** [comment]
@paulocoutinhox 
Can it (you) close?
